File: lib/hecksagain/bluebook/expression/resolver.rb
require "json"
require_relative "../../rendering"

module Hecksagain
  module Bluebook
    module Expression
      class EvaluationError < StandardError; end

      module Resolver
        SIGN_TESTS = %w[positive? negative? zero?].freeze

        # Which Comparison operator each sign test is sugar for, against the
        # literal 0 — declared the same way in Vocabulary::SignTest's
        # compares_via (language/bluebook/vocabulary.bluebook) ; spec/vocabulary_conformance_spec
        # holds the two tables equal.
        SIGN_TEST_OPERATORS = {
          "positive?" => ">",
          "negative?" => "<",
          "zero?"     => "=="
        }.freeze

        # The leaf grammar an expression's dotted/arithmetic side parses into.
        # Which node a string produces is a pure function of the string, so
        # parse() runs once per distinct leaf (folded into Evaluator's own
        # ast_cache — see evaluator.rb) and interpret() reads state/attrs
        # fresh on every call. Only Lookup — the true leaf — ever touches
        # state/attrs.
        IntegerLiteral = Struct.new(:value, keyword_init: true)
        FloatLiteral   = Struct.new(:value, keyword_init: true)
        StringLiteral  = Struct.new(:value, keyword_init: true)
        BoolLiteral    = Struct.new(:value, keyword_init: true)
        # `["active", "suspended"]` — a literal set, the haystack half of
        # an `.include?`. Vocabulary::IncludeHaystack has always ADMITTED
        # Array (and `Evaluator#includes?` has always had an `when Array`
        # arm), but nothing could ever produce one: the resolver had no
        # array literal, so `["a", "b"].include?(x)` fell through to
        # `Lookup` and refused with `cannot resolve "[\"a\", \"b\"]"`.
        # A declared capability with no way to spell it.
        ArrayLiteral   = Struct.new(:elements, keyword_init: true)
        # A plain class, not `Struct.new(keyword_init: true)` — every
        # sibling leaf node here carries at least one field, but this
        # one carries none by nature (a nil literal has no data to
        # hold), and `Struct.new` with zero member names ahead of
        # `keyword_init:` is real, live Ruby-version-dependent
        # behavior: works on 3.3, raises "wrong number of arguments
        # (given 0, expected 1+)" on 3.2 (caught deploying to Lambda's
        # own ruby3.2 runtime). `.new`/`case ... when NilLiteral` below
        # are the only two things this type is ever used for, and a
        # bare class answers both identically.
        NilLiteral     = Class.new
        Addition       = Struct.new(:left, :right, keyword_init: true)
        SignTest       = Struct.new(:operator, :test, :receiver, keyword_init: true)
        Empty          = Struct.new(:receiver, keyword_init: true)
        ToS            = Struct.new(:receiver, keyword_init: true)
        Modulo         = Struct.new(:receiver, :divisor, keyword_init: true)
        Size           = Struct.new(:receiver, keyword_init: true)
        Lookup         = Struct.new(:path, keyword_init: true)

        module_function

        def resolve(expr, state, attrs)
          interpret(parse(expr), state, attrs)
        end

        def parse(expr)
          expr = expr.to_s.strip

          return Size.new(receiver: parse(Regexp.last_match(1))) if expr =~ /\A(.+)\.length\z/

          return IntegerLiteral.new(value: Integer(expr, 10)) if expr.match?(/\A-?\d+\z/)
          return FloatLiteral.new(value: Float(expr))         if expr.match?(/\A-?\d*\.\d+\z/)
          return StringLiteral.new(value: expr[1..-2])        if quoted?(expr)
          return BoolLiteral.new(value: true)                 if expr == "true"
          return BoolLiteral.new(value: false)                if expr == "false"
          return NilLiteral.new                                if expr == "nil"
          elements = array_elements(expr)
          return ArrayLiteral.new(elements: elements.map { |element| parse(element) }) if elements

          arithmetic = split_addition(expr)
          return Addition.new(left: parse(arithmetic[0]), right: parse(arithmetic[1])) if arithmetic

          sign = match_suffix(expr, SIGN_TESTS)
          return sign_test_node(sign) if sign

          return Empty.new(receiver: parse(Regexp.last_match(1))) if expr =~ /\A(.+)\.empty\?\z/
          return ToS.new(receiver: parse(Regexp.last_match(1)))   if expr =~ /\A(.+)\.to_s\z/

          modulo = match_call(expr, ".modulo(")
          return Modulo.new(receiver: parse(modulo[0]), divisor: parse(modulo[1])) if modulo

          return Size.new(receiver: parse(Regexp.last_match(1))) if expr =~ /\A(.+)\.size\z/

          Lookup.new(path: expr)
        end

        def sign_test_node(parts)
          receiver, test = parts
          symbol   = SIGN_TEST_OPERATORS.fetch(test)
          operator = Evaluator::OPERATORS.find { |candidate| candidate.symbol == symbol }
          SignTest.new(operator: operator, test: test, receiver: parse(receiver))
        end

        def interpret(node, state, attrs)
          case node
          when IntegerLiteral, FloatLiteral, StringLiteral, BoolLiteral then node.value
          when ArrayLiteral then node.elements.map { |element| interpret(element, state, attrs) }
          when NilLiteral then nil
          when Addition
            add(interpret(node.left, state, attrs), interpret(node.right, state, attrs))
          when SignTest
            apply_sign_test(node, interpret(node.receiver, state, attrs))
          when Empty
            emptiness_of(interpret(node.receiver, state, attrs))
          when ToS
            string_of(interpret(node.receiver, state, attrs))
          when Modulo
            apply_modulo(interpret(node.receiver, state, attrs), interpret(node.divisor, state, attrs))
          when Size
            size_of(interpret(node.receiver, state, attrs))
          when Lookup
            lookup(node.path, state, attrs)
          end
        end

        # The elements of a bracketed literal, or nil if this isn't one.
        # Splits on TOP-LEVEL commas only — quote-aware and depth-aware,
        # the same discipline `split_addition` already applies, so a
        # nested array or a comma inside a string element stays whole.
        def array_elements(expr)
          return nil unless expr.start_with?("[") && expr.end_with?("]")

          inner = expr[1..-2].strip
          return [] if inner.empty?

          elements = []
          depth = 0
          quote = nil
          current = +""
          inner.each_char do |char|
            if quote
              quote = nil if char == quote
              current << char
              next
            end
            case char
            when '"', "'" then quote = char
            when "[", "(" then depth += 1
            when "]", ")" then depth -= 1
            end
            if char == "," && depth.zero?
              elements << current.strip
              current = +""
            else
              current << char
            end
          end
          elements << current.strip
          elements.reject(&:empty?)
        end

        def split_addition(expr)
          depth = 0
          quote = nil

          expr.each_char.with_index do |char, index|
            if quote
              quote = nil if char == quote
            elsif ['"', "'"].include?(char)
              quote = char
            elsif char == "("
              depth += 1
            elsif char == ")"
              depth -= 1
            elsif char == "+" && depth.zero?
              return [expr[0...index].strip, expr[(index + 1)..].strip]
            end
          end
          nil
        end

        def add(left, right)
          require_number(left, "addition") + require_number(right, "addition")
        end

        def quoted?(expr)
          return false if expr.length < 2

          (expr.start_with?('"') && expr.end_with?('"')) ||
            (expr.start_with?("'") && expr.end_with?("'"))
        end

        # Declared the same way in Vocabulary::SizedType
        # (language/bluebook/vocabulary.bluebook) — spec/vocabulary_conformance_spec
        # holds this equal to the language. Shared by .size and .empty?,
        # which admit the same set for the same reason.
        SIZED_TYPES = %w[Array String Hash].freeze

        def size_of(value)
          return value.size if value.is_a?(Array) || value.is_a?(String) || value.is_a?(Hash)

          raise EvaluationError, "size expects a list or string, got #{describe(value)}"
        end

        def emptiness_of(value)
          return value.empty? if value.is_a?(Array) || value.is_a?(String) || value.is_a?(Hash)

          raise EvaluationError, "empty? expects a list or string, got #{describe(value)}"
        end

        # Declared the same way in Vocabulary::ToStringType
        # (language/bluebook/vocabulary.bluebook) — spec/vocabulary_conformance_spec
        # holds this equal to the language.
        TO_STRING_TYPES = %w[String Integer Float TrueClass FalseClass NilClass].freeze

        def string_of(value)
          case value
          when String                then value
          when Integer, Float        then value.to_s
          when TrueClass, FalseClass then value.to_s
          when NilClass              then ""
          else
            raise EvaluationError, "to_s expects a scalar, got #{describe(value)}"
          end
        end

        def match_suffix(expr, suffixes)
          suffixes.each do |suffix|
            marker = ".#{suffix}"
            return [expr[0...-marker.length], suffix] if expr.end_with?(marker)
          end
          nil
        end

        def apply_sign_test(node, value)
          number = numeric(value)
          raise EvaluationError, "#{node.test} expects a number, got #{describe(value)}" unless number

          Evaluator.apply(node.operator, number, 0)
        end

        def match_call(expr, marker)
          index = expr.rindex(marker)
          return nil unless index && expr.end_with?(")")

          [expr[0...index], expr[(index + marker.length)...-1]]
        end

        def apply_modulo(receiver_value, divisor_value)
          divisor = require_number(divisor_value, "modulo")
          raise EvaluationError, "divided by 0" if divisor.zero?

          require_number(receiver_value, "modulo").to_i % divisor.to_i
        end

        def lookup(expr, state, attrs)
          return fetch(expr, state, attrs) unless expr.include?(".")

          head, *rest = expr.split(".")
          rest.reduce(fetch(head, state, attrs)) do |value, segment|
            break nil unless value.respond_to?(:[])

            value[segment.to_sym] || value[segment]
          end
        end

        def fetch(name, state, attrs)
          key = name.to_sym
          return attrs[key] if attrs.key?(key)
          return state[key] if known?(state, key)

          raise EvaluationError, "cannot resolve #{name.inspect} — no such attribute or argument"
        end

        def known?(state, key)
          return state.key?(key) if state.respond_to?(:key?)

          !state[key].nil?
        end

        def numeric(value)
          value if value.is_a?(Integer) || value.is_a?(Float)
        end

        def require_number(value, operation)
          numeric(value) ||
            raise(EvaluationError, "#{operation} expects a number, got #{describe(value)}")
        end

        def describe(value) = Rendering.describe(value)
      end
    end
  end
end

File: rust/src/bluebook/expression/resolver.rs
use super::evaluator::{apply, Operator, OPERATORS};
use serde_json::{Map, Value};

pub type State = Map<String, Value>;

pub type Eval<T> = Result<T, String>;

const SIGN_TESTS: [&str; 3] = ["positive?", "negative?", "zero?"];

// The leaf grammar an expression's dotted/arithmetic side parses into. Which
// node a string produces is a pure function of the string, so parse() runs
// once per distinct leaf — folded into Evaluator's own AST_CACHE, since a
// leaf string is a substring of some canonical predicate that only gets
// parsed on that predicate's first evaluation (see evaluator.rs) — and
// interpret() reads state/attrs fresh on every call. Only Lookup, the true
// leaf, ever touches state/attrs.
#[derive(Debug)]
pub(crate) enum ResolverNode {
    IntegerLiteral(i64),
    FloatLiteral(f64),
    StringLiteral(String),
    BoolLiteral(bool),
    NilLiteral,
    Addition(Box<ResolverNode>, Box<ResolverNode>),
    SignTest {
        operator: &'static Operator,
        test: &'static str,
        receiver: Box<ResolverNode>,
    },
    Empty(Box<ResolverNode>),
    ToS(Box<ResolverNode>),
    Modulo(Box<ResolverNode>, Box<ResolverNode>),
    Size(Box<ResolverNode>),
    Lookup(String),
}

pub fn resolve_expr(expr: &str, state: &State, attrs: &State) -> Eval<Value> {
    interpret(&parse(expr), state, attrs)
}

pub(crate) fn parse(expr: &str) -> ResolverNode {
    let expr = expr.trim();

    if let Some(field) = expr.strip_suffix(".length") {
        return ResolverNode::Size(Box::new(parse(field.trim())));
    }

    if let Ok(n) = expr.parse::<i64>() {
        return ResolverNode::IntegerLiteral(n);
    }
    if let Ok(f) = expr.parse::<f64>() {
        return ResolverNode::FloatLiteral(f);
    }
    if is_quoted(expr) {
        return ResolverNode::StringLiteral(expr[1..expr.len() - 1].to_string());
    }
    match expr {
        "true" => return ResolverNode::BoolLiteral(true),
        "false" => return ResolverNode::BoolLiteral(false),
        "nil" | "null" => return ResolverNode::NilLiteral,
        _ => {}
    }

    if let Some((left, right)) = split_addition(expr) {
        return ResolverNode::Addition(Box::new(parse(&left)), Box::new(parse(&right)));
    }

    for test in SIGN_TESTS {
        if let Some(receiver) = expr.strip_suffix(&format!(".{}", test)) {
            return sign_test_node(receiver, test);
        }
    }

    if let Some(receiver) = expr.strip_suffix(".empty?") {
        return ResolverNode::Empty(Box::new(parse(receiver.trim())));
    }

    if let Some(receiver) = expr.strip_suffix(".to_s") {
        return ResolverNode::ToS(Box::new(parse(receiver.trim())));
    }

    if let Some((receiver, argument)) = match_call(expr, ".modulo(") {
        return ResolverNode::Modulo(Box::new(parse(&receiver)), Box::new(parse(&argument)));
    }

    if let Some(receiver) = expr.strip_suffix(".size") {
        return ResolverNode::Size(Box::new(parse(receiver.trim())));
    }

    ResolverNode::Lookup(expr.to_string())
}

fn sign_test_node(receiver: &str, test: &'static str) -> ResolverNode {
    let symbol = sign_test_operator(test);
    let operator = OPERATORS
        .iter()
        .find(|candidate| candidate.symbol == symbol)
        .expect("sign_test_operator names a declared Comparison operator");
    ResolverNode::SignTest {
        operator,
        test,
        receiver: Box::new(parse(receiver)),
    }
}

pub(crate) fn interpret(node: &ResolverNode, state: &State, attrs: &State) -> Eval<Value> {
    match node {
        ResolverNode::IntegerLiteral(n) => Ok(Value::from(*n)),
        ResolverNode::FloatLiteral(f) => Ok(Value::from(*f)),
        ResolverNode::StringLiteral(text) => Ok(Value::String(text.clone())),
        ResolverNode::BoolLiteral(flag) => Ok(Value::Bool(*flag)),
        ResolverNode::NilLiteral => Ok(Value::Null),
        ResolverNode::Addition(left, right) => {
            let left = require_number(&interpret(left, state, attrs)?, "addition")?;
            let right = require_number(&interpret(right, state, attrs)?, "addition")?;
            Ok(Value::from(left + right))
        }
        ResolverNode::SignTest {
            operator,
            test,
            receiver,
        } => {
            let value = interpret(receiver, state, attrs)?;
            let number = numeric_value(&value)
                .ok_or_else(|| format!("{} expects a number, got {}", test, describe(&value)))?;
            Ok(Value::Bool(apply(
                operator,
                &Value::from(number),
                &Value::from(0),
            )?))
        }
        ResolverNode::Empty(receiver) => emptiness_of(&interpret(receiver, state, attrs)?),
        ResolverNode::ToS(receiver) => string_of(&interpret(receiver, state, attrs)?),
        ResolverNode::Modulo(receiver, argument) => {
            let divisor = require_number(&interpret(argument, state, attrs)?, "modulo")?;
            if divisor == 0.0 {
                return Err("divided by 0".to_string());
            }
            let left = require_number(&interpret(receiver, state, attrs)?, "modulo")?;
            Ok(Value::from((left as i64).rem_euclid(divisor as i64)))
        }
        ResolverNode::Size(receiver) => size_of(&interpret(receiver, state, attrs)?),
        ResolverNode::Lookup(path) => lookup(path, state, attrs),
    }
}

fn split_addition(expr: &str) -> Option<(String, String)> {
    let mut depth = 0i32;
    let mut quote: Option<char> = None;

    for (index, character) in expr.char_indices() {
        if let Some(open) = quote {
            if character == open {
                quote = None;
            }
            continue;
        }

        match character {
            '\'' | '"' => quote = Some(character),
            '(' => depth += 1,
            ')' => depth -= 1,
            '+' if depth == 0 => {
                return Some((
                    expr[..index].trim().to_string(),
                    expr[index + 1..].trim().to_string(),
                ));
            }
            _ => {}
        }
    }
    None
}

pub fn is_quoted(expr: &str) -> bool {
    let bytes = expr.as_bytes();
    if bytes.len() < 2 {
        return false;
    }
    let (first, last) = (bytes[0], bytes[bytes.len() - 1]);
    (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'')
}

fn size_of(value: &Value) -> Eval<Value> {
    match value {
        Value::Array(items) => Ok(Value::from(items.len() as i64)),
        Value::String(text) => Ok(Value::from(text.chars().count() as i64)),
        Value::Object(map) => Ok(Value::from(map.len() as i64)),
        other => Err(format!(
            "size expects a list or string, got {}",
            describe(other)
        )),
    }
}

fn emptiness_of(value: &Value) -> Eval<Value> {
    match value {
        Value::Array(items) => Ok(Value::Bool(items.is_empty())),
        Value::String(text) => Ok(Value::Bool(text.is_empty())),
        Value::Object(map) => Ok(Value::Bool(map.is_empty())),
        other => Err(format!(
            "empty? expects a list or string, got {}",
            describe(other)
        )),
    }
}

fn string_of(value: &Value) -> Eval<Value> {
    Ok(Value::String(match value {
        Value::String(text) => text.clone(),
        Value::Null => String::new(),
        Value::Bool(flag) => flag.to_string(),
        Value::Number(n) => match n.as_i64() {
            Some(i) => i.to_string(),
            None => format_float(n.as_f64().unwrap_or(0.0)),
        },
        other => return Err(format!("to_s expects a scalar, got {}", describe(other))),
    }))
}

// Which Comparison operator each sign test is sugar for, against the literal
// 0 — declared the same way in Vocabulary::SignTest's compares_via
// (language/bluebook.bluebook) ; spec/vocabulary_conformance_spec holds the
// two tables equal.
fn sign_test_operator(test: &str) -> &'static str {
    match test {
        "positive?" => ">",
        "negative?" => "<",
        _ => "==",
    }
}

pub fn match_call(expr: &str, marker: &str) -> Option<(String, String)> {
    if !expr.ends_with(')') {
        return None;
    }
    let index = expr.rfind(marker)?;
    Some((
        expr[..index].to_string(),
        expr[index + marker.len()..expr.len() - 1].to_string(),
    ))
}

fn lookup(expr: &str, state: &State, attrs: &State) -> Eval<Value> {
    if !expr.contains('.') {
        return fetch(expr, state, attrs);
    }

    let mut segments = expr.split('.');
    let head = segments.next().unwrap_or_default();
    let mut current = fetch(head, state, attrs)?;

    for segment in segments {
        current = match current {
            Value::Object(map) => map.get(segment).cloned().unwrap_or(Value::Null),
            _ => return Ok(Value::Null),
        };
    }
    Ok(current)
}

fn fetch(name: &str, state: &State, attrs: &State) -> Eval<Value> {
    if let Some(value) = attrs.get(name) {
        return Ok(value.clone());
    }
    if let Some(value) = state.get(name) {
        return Ok(value.clone());
    }
    Err(format!(
        "cannot resolve {:?} \u{2014} no such attribute or argument",
        name
    ))
}

pub fn numeric_value(value: &Value) -> Option<f64> {
    match value {
        Value::Number(n) => n.as_f64(),
        _ => None,
    }
}

pub fn require_number(value: &Value, operation: &str) -> Eval<f64> {
    numeric_value(value)
        .ok_or_else(|| format!("{} expects a number, got {}", operation, describe(value)))
}

pub fn describe(value: &Value) -> String {
    match value {
        Value::Null => "nil".to_string(),
        Value::String(text) => format!("{:?}", text),
        Value::Number(n) => match n.as_i64() {
            Some(i) => i.to_string(),
            None => format_float(n.as_f64().unwrap_or(0.0)),
        },
        other => other.to_string(),
    }
}

pub fn class_of(value: &Value) -> String {
    match value {
        Value::Null => "nil".to_string(),
        Value::Bool(true) => "TrueClass".to_string(),
        Value::Bool(false) => "FalseClass".to_string(),
        Value::String(_) => "String".to_string(),
        Value::Array(_) => "Array".to_string(),
        Value::Object(_) => "Hash".to_string(),
        Value::Number(n) => {
            if n.is_i64() || n.is_u64() {
                "Integer".to_string()
            } else {
                "Float".to_string()
            }
        }
    }
}

fn format_float(value: f64) -> String {
    if value.fract() == 0.0 {
        format!("{:.1}", value)
    } else {
        value.to_string()
    }
}
